src/NuGet.CommandLine/Commands/ListCommand.cs

```diff
@@ -56,35 +115,23 @@ public ICollection<string> Source
                     Protocol.Core.v2.FactoryExtensionsV2.GetCoreV2(Repository.Provider),
                     Protocol.Core.v3.FactoryExtensionsV2.GetCoreV3(Repository.Provider)));
 
-            var resourceTasks = new List<Task<SimpleSearchResource>>();
+            var listCommandResourceTasks = new List<Task<ListCommandResource>>();
+
             foreach (var source in packageSources)
             {
                 var sourceRepository = sourceRepositoryProvider.CreateRepository(source);
-                resourceTasks.Add(sourceRepository.GetResourceAsync<SimpleSearchResource>());
+                listCommandResourceTasks.Add(sourceRepository.GetResourceAsync<ListCommandResource>());
             }
-            var resources = await Task.WhenAll(resourceTasks);
-
-            var searchTerm = Arguments.FirstOrDefault();
-            var resultTasks = resources.Where(r => r != null)
-                .Select(r => r.Search(
-                    searchTerm,
-                    new SearchFilter(Enumerable.Empty<string>(), Prerelease, includeDelisted: false),
-                    skip: 0,
-                    take: string.IsNullOrEmpty(searchTerm) ? UnfilteredPageSize : FilteredPageSize,
-                    cancellationToken: CancellationToken.None));
-
-            var results = (await Task.WhenAll(resultTasks)).SelectMany(feedResult => feedResult);
-
-            //if (AllVersions)
-            //{
-            //    results = results.SelectMany(package => package.AllVersions.Select(version =>
-            //                new SimpleSearchMetadata(
-            //                    new Packaging.Core.PackageIdentity(package.Identity.Id, version), 
-            //                    package.Description, 
-            //                    Enumerable.Empty<NuGetVersion>())));
-            //}
-
-            return results.OrderBy(s => s.Identity.Id, StringComparer.Ordinal);
+            var listCommandResources = await Task.WhenAll(listCommandResourceTasks);
+
+            var listEndpoints = new List<string>();
+            foreach(var listCommandResource in listCommandResources)
+            {
+                var listEndpoint = await listCommandResource.GetListEndpointAsync(CancellationToken.None);
+                listEndpoints.Add(listEndpoint);
```

Comment:
listEndpoint can be null here if a v3 source does not contain the correct resource in index.json. In that case I would expect that we just ignore the source and keep it out of the list. If null gets added it looks like it will throw later on.